File: 실습/4장/04_constructor_init/Main.java
public class Main {

    static class User {
        private final String name;
        private int age;

        public User(String name) {
            this.name = name;
            System.out.println(name + " 사용자 생성");
        }

        public User(String name, int age) {
            this(name);
            setAge(age);
        }

        public String getName() {
            return name;
        }

        public int getAge() {
            return age;
        }

        public void setAge(int age) {
            if (age < 0) {
                throw new IllegalArgumentException(
                        "나이는 음수일 수 없습니다."
                );
            }

            this.age = age;
        }
    }

    public static void main(String[] args) {
        User user = new User("철수", 20);

        System.out.println(user.getName());
        System.out.println(user.getAge());
    }
}

File: 실습/4장/04_constructor_init/Main.kt
class User(
    val name: String
) {
    var age: Int = 0
        set(value) {
            require(value >= 0) {
                "나이는 음수일 수 없습니다."
            }
            field = value
        }

    init {
        println("$name 사용자 생성")
    }

    constructor(name: String, age: Int) : this(name) {
        this.age = age
    }
}

fun main() {
    val user = User("철수", 20)

    println(user.name)
    println(user.age)
}